# Data import and descriptive analysis 

In [1]:
# Import libraries 

import pandas as pd 
import numpy as np
import os

In [2]:
path  = r'/Users/lupeguevara/Documents/DOE Demographic Snapshot'

In [3]:
path

'/Users/lupeguevara/Documents/DOE Demographic Snapshot'

In [4]:
# Import dataset 
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '2013_-_2018_Demographic_Snapshot_School_20240215.csv'), index_col = False)

# Data wrangling 

In [5]:
df.head()

DBN                School Name     Year  Total Enrollment  \
0  01M015  P.S. 015 Roberto Clemente  2013-14               190   
1  01M015  P.S. 015 Roberto Clemente  2014-15               183   
2  01M015  P.S. 015 Roberto Clemente  2015-16               176   
3  01M015  P.S. 015 Roberto Clemente  2016-17               178   
4  01M015  P.S. 015 Roberto Clemente  2017-18               190   

   Grade PK (Half Day & Full Day)  Grade K  Grade 1  Grade 2  Grade 3  \
0                              26       39       39       21       16   
1                              18       27       47       31       19   
2                              14       32       33       39       23   
3                              17       28       33       27       31   
4                              17       28       32       33       23   

   Grade 4  ...  % Multiple Race Categories Not Represented  # White  % White  \
0       26  ...                                         1.1        3      1.6   
1       17  ...                                         0.5        2      1.1   
2       17  ...                                         1.7        2      1.1   
3       24  ...                                         2.2        4      2.2   
4       31  ...                                         1.1        6      3.2   

   # Students with Disabilities  % Students with Disabilities  \
0                            65                          34.2   
1                            64                          35.0   
2                            60                          34.1   
3                            51                          28.7   
4                            45                          23.7   

   # English Language Learners  % English Language Learners  # Poverty  \
0                           19                         10.0        171   
1                           17                          9.3        169   
2                           16                          9.1        149   
3                           12                          6.7        152   
4                            8                          4.2        161   

   % Poverty  Economic Need Index  
0       90.0              No Data  
1       92.3                93.5%  
2       84.7                89.6%  
3       85.4                89.2%  
4       84.7                89.0%  

[5 rows x 39 columns]

In [6]:
df.tail()

DBN                        School Name     Year  Total Enrollment  \
8967  84X730  Bronx Charter School for the Arts  2013-14               319   
8968  84X730  Bronx Charter School for the Arts  2014-15               316   
8969  84X730  Bronx Charter School for the Arts  2015-16               323   
8970  84X730  Bronx Charter School for the Arts  2016-17               320   
8971  84X730  Bronx Charter School for the Arts  2017-18               314   

      Grade PK (Half Day & Full Day)  Grade K  Grade 1  Grade 2  Grade 3  \
8967                               0       52       54       56       56   
8968                               0       53       55       56       53   
8969                               0       51       57       55       55   
8970                               0       53       53       55       52   
8971                               0       50       51       54       51   

      Grade 4  ...  % Multiple Race Categories Not Represented  # White  \
8967       53  ...                                         1.9        3   
8968       47  ...                                         1.6        2   
8969       57  ...                                         1.9        5   
8970       53  ...                                         0.9        3   
8971       52  ...                                         1.0        1   

      % White  # Students with Disabilities  % Students with Disabilities  \
8967      0.9                            57                          17.9   
8968      0.6                            53                          16.8   
8969      1.5                            61                          18.9   
8970      0.9                            67                          20.9   
8971      0.3                            63                          20.1   

      # English Language Learners  % English Language Learners  # Poverty  \
8967                           27                          8.5        222   
8968                           34                         10.8        284   
8969                           49                         15.2        268   
8970                           51                         15.9        235   
8971                           57                         18.2        253   

      % Poverty  Economic Need Index  
8967       69.6              No Data  
8968       89.9                82.2%  
8969       83.0                80.6%  
8970       73.4                83.5%  
8971       80.6                88.3%  

[5 rows x 39 columns]

In [7]:
df.shape

(8972, 39)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8972 entries, 0 to 8971
Data columns (total 39 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   DBN                                         8972 non-null   object 
 1   School Name                                 8972 non-null   object 
 2   Year                                        8972 non-null   object 
 3   Total Enrollment                            8972 non-null   int64  
 4   Grade PK (Half Day & Full Day)              8972 non-null   int64  
 5   Grade K                                     8972 non-null   int64  
 6   Grade 1                                     8972 non-null   int64  
 7   Grade 2                                     8972 non-null   int64  
 8   Grade 3                                     8972 non-null   int64  
 9   Grade 4                                     8972 non-null   int64  
 10  Grade 5     

In [9]:
# Removing the limit to columns and rows displayed so I can see all the rows and columns going forward
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [10]:
# Looking for basic stats for consistency check
df.describe()

Total Enrollment  Grade PK (Half Day & Full Day)      Grade K  \
count       8972.000000                     8972.000000  8972.000000   
mean         598.226148                       15.635533    46.608671   
std          485.684603                       36.119885    54.561859   
min           12.000000                        0.000000     0.000000   
25%          323.000000                        0.000000     0.000000   
50%          484.000000                        0.000000    34.000000   
75%          704.000000                       31.000000    83.000000   
max         5838.000000                     1158.000000   406.000000   

           Grade 1      Grade 2      Grade 3      Grade 4      Grade 5  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     48.398128    47.649131    46.430227    45.044026    43.910945   
std      57.195199    57.255600    57.788614    57.712646    56.348207   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%      35.000000    31.500000    24.000000     0.000000     0.000000   
75%      86.000000    86.000000    84.000000    81.000000    80.000000   
max     383.000000   349.000000   373.000000   376.000000   351.000000   

           Grade 6      Grade 7      Grade 8      Grade 9     Grade 10  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     42.599643    41.859563    41.536781    50.853879    49.713553   
std      87.889778    89.773353    91.209942   135.137015   151.644252   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%      63.000000    60.000000    58.000000    69.000000    70.000000   
max     771.000000   796.000000   839.000000  1551.000000  3832.000000   

          Grade 11     Grade 12     # Female     % Female       # Male  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     39.446946    38.539122   290.397793    48.342499   307.820218   
std     108.051316   105.211557   242.665721     9.572231   257.627423   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000   145.000000    46.200000   162.000000   
50%       0.000000     0.000000   237.000000    48.800000   251.000000   
75%      52.000000    48.000000   353.000000    51.300000   369.000000   
max    1450.000000  1380.000000  2360.000000   100.000000  3478.000000   

            % Male      # Asian      % Asian      # Black      % Black  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     51.656498    94.344182    11.043792   161.841953    32.952786   
std       9.572352   217.781445    16.496850   172.300146    27.507728   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%      48.700000     5.000000     1.200000    44.000000     9.000000   
50%      51.200000    16.000000     3.700000   111.000000    26.300000   
75%      53.800000    74.000000    12.900000   208.000000    52.700000   
max     100.000000  3576.000000    94.700000  1481.000000    98.800000   

        # Hispanic   % Hispanic  # Multiple Race Categories Not Represented  \
count  8972.000000  8972.000000                                 8972.000000   
mean    242.073005    41.932055                                   12.493981   
std     235.650938    25.225109                                   19.403410   
min       1.000000     0.400000                                    0.000000   
25%      84.000000    19.400000                                    3.000000   
50%     175.000000    39.000000                                    7.000000   
75%     312.000000    62.700000                                   14.000000   
max    2005.000000   100.000000                           

#### Total enrollment:
The mean total enrollment across the dataset is approximately 598 students per institution.
The standard deviation of total enrollment is relatively high at 485, indicating considerable variability in enrollment sizes among institutions.
The smallest institution has 12 students enrolled, while the largest institution has 5838 students.

#### Grade Distribution (PK to Grade 7):

The mean enrollment for different grades varies, with the highest mean for Grade K (Kindergarten) at approximately 46 students.
There's a substantial standard deviation for each grade level, indicating variability in enrollment sizes across different grades.
The minimum and maximum enrollment for each grade level show a wide range, reflecting diverse student populations across institutions.

In [11]:
df.head(100)

DBN                            School Name     Year  Total Enrollment  \
0   01M015              P.S. 015 Roberto Clemente  2013-14               190   
1   01M015              P.S. 015 Roberto Clemente  2014-15               183   
2   01M015              P.S. 015 Roberto Clemente  2015-16               176   
3   01M015              P.S. 015 Roberto Clemente  2016-17               178   
4   01M015              P.S. 015 Roberto Clemente  2017-18               190   
5   01M019                    P.S. 019 Asher Levy  2013-14               285   
6   01M019                    P.S. 019 Asher Levy  2014-15               270   
7   01M019                    P.S. 019 Asher Levy  2015-16               270   
8   01M019                    P.S. 019 Asher Levy  2016-17               271   
9   01M019                    P.S. 019 Asher Levy  2017-18               257   
10  01M020                   P.S. 020 Anna Silver  2013-14               631   
11  01M020                   P.S. 020 Anna Silver  2014-15               633   
12  01M020                   P.S. 020 Anna Silver  2015-16               581   
13  01M020                   P.S. 020 Anna Silver  2016-17               540   
14  01M020                   P.S. 020 Anna Silver  2017-18               497   
15  01M034         P.S. 034 Franklin D. Roosevelt  2013-14               393   
16  01M034         P.S. 034 Franklin D. Roosevelt  2014-15               395   
17  01M034         P.S. 034 Franklin D. Roosevelt  2015-16               394   
18  01M034         P.S. 034 Franklin D. Roosevelt  2016-17               350   
19  01M034         P.S. 034 Franklin D. Roosevelt  2017-18               333   
20  01M063              The STAR Academy - P.S.63  2013-14               179   
21  01M063              The STAR Academy - P.S.63  2014-15               199   
22  01M063              The STAR Academy - P.S.63  2015-16               203   
23  01M063              The STAR Academy - P.S.63  2016-17               200   
24  01M063              The STAR Academy - P.S.63  2017-18               203   
25  01M064                  P.S. 064 Robert Simon  2013-14               287   
26  01M064                  P.S. 064 Robert Simon  2014-15               282   
27  01M064                  P.S. 064 Robert Simon  2015-16               258   
28  01M064                  P.S. 064 Robert Simon  2016-17               247   
29  01M064                  P.S. 064 Robert Simon  2017-18               245   
30  01M110          P.S. 110 Florence Nightingale  2013-14               416   
31  01M110          P.S. 110 Florence Nightingale  2014-15               419   
32  01M110          P.S. 110 Florence Nightingale  2015-16               394   
33  01M110          P.S. 110 Florence Nightingale  2016-17               383   
34  01M110          P.S. 110 Florence Nightingale  2017-18               379   
35  01M134               P.S. 134 Henrietta Szold  2013-14               318   
36  01M134               P.S. 134 Henrietta Szold  2014-15               255   
37  01M134               P.S. 134 Henrietta Szold  2015-16               263   
38  01M134               P.S. 134 Henrietta Szold  2016-17               354   
39  01M134               P.S. 134 Henrietta Szold  2017-18               297   
40  01M140                 P.S. 140 Nathan Straus  2013-14               393   
41  01M140                 P.S. 140 Nathan Straus  2014-15               380   
42  01M140                 P.S. 140 Nathan Straus  2015-16               376   
43  01M140                 P.S. 140 Nathan Straus  2016-17               382   
44  01M140                 P.S. 140 Nathan Straus  2017-18               371   
45  01M142                 P.S. 142 Amalia Castro  2013-14               384   
46  01M142                 P.S. 142 Amalia Castro  2014-15               386   
47  01M142                 P.S. 142 Amalia Castro  2015-16               354   
48  01M142                 P.S. 142 Amalia Castro  2016-17               356   
49  01M

#### Some variables have both number (#) and percentage (%) form in the dataset. By dropping the number form of the variables, you maintain the essential information while streamlining the dataset and making it more manageable for analysis.  

In [12]:
# Removing unnecessary columns
df=df.drop(columns = ['# Female', '# Male', '# Asian', '# Black', '# Hispanic', '# Multiple Race Categories Not Represented', '# White', '# Students with Disabilities', '# English Language Learners', '# Poverty'])

In [13]:
# Checking that 10 columns were dropped
df.shape

(8972, 29)

In [14]:
# Changing name of column from 'DBN' to 'District Borough Number'
df.rename(columns = {'DBN' : 'District Borough Number'}, inplace = True)

In [15]:
# Verify changes to renaming column
df.head()

District Borough Number                School Name     Year  \
0                  01M015  P.S. 015 Roberto Clemente  2013-14   
1                  01M015  P.S. 015 Roberto Clemente  2014-15   
2                  01M015  P.S. 015 Roberto Clemente  2015-16   
3                  01M015  P.S. 015 Roberto Clemente  2016-17   
4                  01M015  P.S. 015 Roberto Clemente  2017-18   

   Total Enrollment  Grade PK (Half Day & Full Day)  Grade K  Grade 1  \
0               190                              26       39       39   
1               183                              18       27       47   
2               176                              14       32       33   
3               178                              17       28       33   
4               190                              17       28       32   

   Grade 2  Grade 3  Grade 4  Grade 5  Grade 6  Grade 7  Grade 8  Grade 9  \
0       21       16       26       23        0        0        0        0   
1       31       19       17       24        0        0        0        0   
2       39       23       17       18        0        0        0        0   
3       27       31       24       18        0        0        0        0   
4       33       23       31       26        0        0        0        0   

   Grade 10  Grade 11  Grade 12  % Female  % Male  % Asian  % Black  \
0         0         0         0      48.9    51.1      4.7     37.9   
1         0         0         0      45.9    54.1      4.4     35.5   
2         0         0         0      47.2    52.8      5.1     32.4   
3         0         0         0      46.6    53.4      7.9     28.7   
4         0         0         0      52.1    47.9     10.5     27.4   

   % Hispanic  % Multiple Race Categories Not Represented  % White  \
0        54.7                                         1.1      1.6   
1        58.5                                         0.5      1.1   
2        59.7                                         1.7      1.1   
3        59.0                                         2.2      2.2   
4        57.9                                         1.1      3.2   

   % Students with Disabilities  % English Language Learners  % Poverty  \
0                          34.2                         10.0       90.0   
1                          35.0                          9.3       92.3   
2                          34.1                          9.1       84.7   
3                          28.7                          6.7       85.4   
4                          23.7                          4.2       84.7   

  Economic Need Index  
0             No Data  
1               93.5%  
2               89.6%  
3               89.2%  
4               89.0%

# Data Consistency Check

In [16]:
# Looking for basic stats for consistency check
df.describe()

Total Enrollment  Grade PK (Half Day & Full Day)      Grade K  \
count       8972.000000                     8972.000000  8972.000000   
mean         598.226148                       15.635533    46.608671   
std          485.684603                       36.119885    54.561859   
min           12.000000                        0.000000     0.000000   
25%          323.000000                        0.000000     0.000000   
50%          484.000000                        0.000000    34.000000   
75%          704.000000                       31.000000    83.000000   
max         5838.000000                     1158.000000   406.000000   

           Grade 1      Grade 2      Grade 3      Grade 4      Grade 5  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     48.398128    47.649131    46.430227    45.044026    43.910945   
std      57.195199    57.255600    57.788614    57.712646    56.348207   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%      35.000000    31.500000    24.000000     0.000000     0.000000   
75%      86.000000    86.000000    84.000000    81.000000    80.000000   
max     383.000000   349.000000   373.000000   376.000000   351.000000   

           Grade 6      Grade 7      Grade 8      Grade 9     Grade 10  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     42.599643    41.859563    41.536781    50.853879    49.713553   
std      87.889778    89.773353    91.209942   135.137015   151.644252   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%      63.000000    60.000000    58.000000    69.000000    70.000000   
max     771.000000   796.000000   839.000000  1551.000000  3832.000000   

          Grade 11     Grade 12     % Female       % Male      % Asian  \
count  8972.000000  8972.000000  8972.000000  8972.000000  8972.000000   
mean     39.446946    38.539122    48.342499    51.656498    11.043792   
std     108.051316   105.211557     9.572231     9.572352    16.496850   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000    46.200000    48.700000     1.200000   
50%       0.000000     0.000000    48.800000    51.200000     3.700000   
75%      52.000000    48.000000    51.300000    53.800000    12.900000   
max    1450.000000  1380.000000   100.000000   100.000000    94.700000   

           % Black   % Hispanic  % Multiple Race Categories Not Represented  \
count  8972.000000  8972.000000                                 8972.000000   
mean     32.952786    41.932055                                    2.129848   
std      27.507728    25.225109                                    2.480443   
min       0.000000     0.400000                                    0.000000   
25%       9.000000    19.400000                                    0.700000   
50%      26.300000    39.000000                                    1.400000   
75%      52.700000    62.700000                                    2.600000   
max      98.800000   100.000000                                   31.200000   

           % White  % Students with Disabilities  % English Language Learners  \
count  8972.000000                    8972.00000                  8972.000000   
mean     11.942142                      22.03765                    13.407769   
std      18.287075                      15.90465                    14.227278   
min       0.000000                       0.00000                     0.000000   
25%       1.200000                      14.90000                     4.100000   
50%       2.900000                      19.40000                     9.200000   
75%      13.900000                      24.50000                    17.800000 

In [17]:
# Investigating Missing Values
df.isnull().sum()

District Borough Number                       0
School Name                                   0
Year                                          0
Total Enrollment                              0
Grade PK (Half Day & Full Day)                0
Grade K                                       0
Grade 1                                       0
Grade 2                                       0
Grade 3                                       0
Grade 4                                       0
Grade 5                                       0
Grade 6                                       0
Grade 7                                       0
Grade 8                                       0
Grade 9                                       0
Grade 10                                      0
Grade 11                                      0
Grade 12                                      0
% Female                                      0
% Male                                        0
% Asian                                 

#### There are no missing values in the dataset to remove. It has been observed that the variable 'Economic Need Index' contains values that says No date instead of NaN. No changes are needed. 

In [18]:
# Looking for mixed data types
for col in df.columns.tolist():
    weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df[weird]) > 0:
        print (col)

#### There is no mixed data 

In [19]:
# Looking for duplicate records
df_dups = df[df.duplicated()]

In [20]:
df_dups

Empty DataFrame
Columns: [District Borough Number, School Name, Year, Total Enrollment, Grade PK (Half Day & Full Day), Grade K, Grade 1, Grade 2, Grade 3, Grade 4, Grade 5, Grade 6, Grade 7, Grade 8, Grade 9, Grade 10, Grade 11, Grade 12, % Female, % Male, % Asian, % Black, % Hispanic, % Multiple Race Categories Not Represented, % White, % Students with Disabilities, % English Language Learners, % Poverty, Economic Need Index]
Index: []

#### There are no duplicate records

In [21]:
# Next will look for inconsistent spellings or entry formats

In [22]:
# Look at unique values in specific columns
unique_values = df['District Borough Number'].unique()

In [23]:
# Print unique values
print(unique_values)

['01M015' '01M019' '01M020' ... '84X717' '84X718' '84X730']


In [24]:
unique_values = df['School Name'].unique()

In [25]:
# Print unique values
print(unique_values)

['P.S. 015 Roberto Clemente' 'P.S. 019 Asher Levy' 'P.S. 020 Anna Silver'
 ... 'Icahn Charter School' 'Bronx Charter School for Better Learning'
 'Bronx Charter School for the Arts']


#### There do not appear to be any inconsistencies in spellings or entry formats.

In [26]:
# Exporting data frame as csv

df.to_csv(os.path.join(path, '02 Data','Prepared Data', 'school_snapshot_clean.csv'))